Mempool API changes for Alonzo.

nc6 opened the pull request:
We now expose two separate functions in the mempool:

- `validateTx`, which takes a transaction and returns a `TxInBlock`.
- `reapplyTx`, which applies a `TxInBlock` to a new ledger state.

nc6 committed:
Mempool API changes for Alonzo.

We now expose three separate functions in the mempool:

- `applyTx`, which takes a transaction and returns a `TxInBlock`.
- `applyTxInBlock`, which applies a `TxInBlock` to a new ledger state.
- `extractTx`, which returns the underlying transaction from a
  `TxInBlock`.

This allows us to reintroduce API instances for Alonzo.

Currently the `ApplyTx` instance is undefined. CAD-2880 and CAD-2881
detail the implementation task.

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Mempool.hs:
```diff
@@ -190,9 +261,14 @@ overNewEpochState ::
+              }
+        }
+  where
+    (>$<) = flip (<$>)
```
:smile:

nfrisby replied:
There's already `Data.Functor.<&>` if you're keen on it https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-Functor.html#v:-60--38--62-

nc6 replied:
Aha, I knew it lived somewhere...

JaredCorduan reviewed (approved):
looks good to me!

nc6 committed:
Do not serialise full script in errors.

The `Phase1ScriptWitnessNotValidating` constructor previously wrapped
the full script, rather than the hash as is the case for all other
constructors. This would break things, since scripts cannot be
deserialised without annotation, and also would entail potentially very
large errors.

Instead, use the script hash as in every other case.

nfrisby commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Mempool.hs:
```diff
@@ -189,9 +263,14 @@ overNewEpochState ::
-      { LedgerState.nesEs =
-          (LedgerState.nesEs st) {LedgerState.esLState = res}
-      }
+  f (mkMempoolState st)
```
Minor: Unless I'm thinking too quickly here, you only need `Functor f`, not `Applicative f`. (IE `overNewEpochState` is just a "lens", not a "traversal".)
